src/CodeFixes/CSharp/CodeFixes/DeclareAsNullableCodeFixProvider.cs

```diff
@@ -85,4 +90,42 @@ private static void TryRegisterCodeFix(CodeFixContext context, Diagnostic diagno
 
         context.RegisterCodeFix(codeAction, diagnostic);
     }
+
+    private static void TryRegisterCodeFixForCast(CodeFixContext context, Diagnostic diagnostic, TypeSyntax type)
+    {
+        if (type.IsKind(SyntaxKind.NullableType))
+            return;
+
+        CodeAction codeAction = CodeAction.Create(
+            "Declare as nullable",
+            async ct =>
+            {
+                NullableTypeSyntax newType = SyntaxFactory.NullableType(type.WithoutTrivia()).WithTriviaFrom(type);
+                Document changedDocument = await context.Document.ReplaceNodeAsync(type, newType, ct).ConfigureAwait(false);
+
+                // This could be in a variable declaration, so grab the new syntax root and find the newly-replaced type node
+                SyntaxNode root = await changedDocument.GetSyntaxRootAsync(ct).ConfigureAwait(false);
+                SyntaxNode insertedNewType = root.FindNode(type.Span);
+
+                // Try finding a surrounding variable declaration whose type we also have to change
+                if (insertedNewType.AncestorsAndSelf().FirstOrDefault(a => a.IsKind(SyntaxKind.EqualsValueClause)) is EqualsValueClauseSyntax equalsValueClause)
+                {
+                    ExpressionSyntax expression = equalsValueClause.Value;
+
+                    if (equalsValueClause.IsParentKind(SyntaxKind.VariableDeclarator)
+                        && equalsValueClause.Parent.Parent is VariableDeclarationSyntax variableDeclaration
+                        && variableDeclaration.Variables.Count == 1
+                        && !variableDeclaration.Type.IsKind(SyntaxKind.NullableType)
+                        && variableDeclaration.Type is not IdentifierNameSyntax { Identifier.Text: "var" })
+                    {
+                        NullableTypeSyntax newDeclarationType = SyntaxFactory.NullableType(variableDeclaration.Type.WithoutTrivia()).WithTriviaFrom(variableDeclaration.Type);
+                        changedDocument = await changedDocument.ReplaceNodeAsync(variableDeclaration.Type, newDeclarationType, ct).ConfigureAwait(false);
```

Comment:
Instead of updating document twice, I would rather use something similar like:

```cs
VariableDeclarationSyntax newVariableDeclaration = variableDeclaration
    .ReplaceNode(type, newType)
    .WithType(newDeclarationType);
```

Reply:
That's ... indeed much simpler. The solution with finding the correct place in the changed document _did_ feel weird and I felt there had to be a better solution. Although it only works in this case because the syntax nodes are practically neighbors, I guess. If the code fix would have to change, say, a field declaration somewhere else, there'd be no way around this dance, right?

Reply:
Updating multiple nodes can be tricky. And there many approaches. You can use SyntaxRewriter, method SyntaxNode.ReplaceNodes, SyntaxAnnotation to find a node in a new tree. If you are replacing members or statements you can use its index to find it in a collection etc.